pkg/codegen/go/gen.go

```diff
@@ -468,6 +508,100 @@ func genInputMethods(w io.Writer, name, receiverType, elementType string, ptrMet
 	}
 }
 
+func (pkg *pkgContext) genEnum(w io.Writer, enum *schema.EnumType) {
+	pkg.genEnumType(w, pkg.tokenToEnum(enum.Token), enum)
+}
+
+func (pkg *pkgContext) genEnumType(w io.Writer, name string, enumType *schema.EnumType) {
+	mod := pkg.tokenToPackage(enumType.Token)
+	modPkg, ok := pkg.packages[mod]
+	contract.Assert(ok)
+	printCommentWithDeprecationMessage(w, enumType.Comment, "", false)
+	elementType := pkg.enumElementType(enumType.ElementType, false)
+	fmt.Fprintf(w, "type %s %s\n\n", name, elementType)
+
+	fmt.Fprintln(w, "const (")
+	for _, e := range enumType.Elements {
+		printCommentWithDeprecationMessage(w, e.Comment, e.DeprecationMessage, true)
+		var elementName = e.Name
+		// Add the type to the name to disambiguate constants used for enum values
+		if e.Name == "" {
+			elementName = fmt.Sprintf("%v", e.Value)
+		}
+		enumName := makeSafeEnumName(elementName)
+		if strings.Contains(enumName, "_") {
+			enumName = fmt.Sprintf("_%s", enumName)
+		}
+		e.Name = name + enumName
+		contract.Assertf(!modPkg.names.has(e.Name), "Name collision for enum constant: %s for %s",
+			e.Name, enumType.Token)
+		switch reflect.TypeOf(e.Value).Kind() {
+		case reflect.String:
+			fmt.Fprintf(w, "%s = %s(\"%v\")\n", e.Name, name, e.Value)
+		default:
+			fmt.Fprintf(w, "%s = %s(%v)\n", e.Name, name, e.Value)
+		}
+	}
+	fmt.Fprintln(w, ")")
+	inputType := pkg.inputType(enumType, false)
+	contract.Assertf(name == inputType,
+		"expect inputType (%s) for enums to be the same as enum type (%s)", inputType, enumType)
+	pkg.genEnumInputFuncs(w, name, enumType, elementType, inputType)
+}
+
+func (pkg *pkgContext) enumElementType(t schema.Type, optional bool) string {
+	suffix := ""
+	if optional {
+		suffix = "Ptr"
+	}
+	switch t {
+	case schema.BoolType:
+		return "pulumi.Bool" + suffix
+	case schema.IntType:
+		return "pulumi.Int" + suffix
+	case schema.NumberType:
+		return "pulumi.Float64" + suffix
+	case schema.StringType:
+		return "pulumi.String" + suffix
+	default:
+		// We only expect to support the above element types for enums
+		panic(t)
+	}
+}
+
+func makeSafeEnumName(name string) string {
+	return makeValidIdentifier(Title(name))
+}
+
+func (pkg *pkgContext) genEnumInputFuncs(w io.Writer, typeName string, enum *schema.EnumType, elementType, inputType string) {
+	fmt.Fprintln(w)
+	asFuncName := Title(strings.Replace(elementType, "pulumi.", "", -1))
+	fmt.Fprintf(w, "func (%s) ElementType() reflect.Type {\n", typeName)
+	fmt.Fprintf(w, "return reflect.TypeOf((*%s)(nil)).Elem()\n", elementType)
+	fmt.Fprintln(w, "}")
+	fmt.Fprintln(w)
+
+	fmt.Fprintf(w, "func (e %[1]s) To%[2]sOutput() %[3]sOutput {\n", typeName, asFuncName, elementType)
```

Comment:
Nit: No need for `[1]`, `[2]`, etc. if we're not repeating the same element. 
```suggestion
	fmt.Fprintf(w, "func (e %s) To%sOutput() %sOutput {\n", typeName, asFuncName, elementType)
```